### PR TITLE
Model Generator Source Paths Should Allow for Customization

### PR DESCRIPTION
**Pull request by spencerneste**
### Motivation / Background

This Pull Request has been created because https://github.com/rails/rails/pull/13972 was closed. It has the same content as the mentioned PR, with the addition of a test that would fail without the new code, and an instance variable initialization to silence a warning. 

### Detail

This Pull Request changes which source paths to consider when using the rails model generator. More detail can be found here: https://github.com/rails/rails/pull/13972#issue-27191554

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by spencerneste** [comment]
@rafaelfranca Based on https://github.com/rails/rails/pull/13972#issuecomment-406074086 , mind taking a look?

> This PR need tests to be merged. If anyone wants to pick it up and open a new PR feel free to do. I'm closing because it is an incomplete old PR. Thank you so much for the contribution.

**Comment by zzak** [comment]
Since this patch is almost the same as before (except with tests), I think it's good to add the original author as well to the commit.

You can do this by adding this to the bottom of your commit message:

```

Co-authored-by: Author Name <email>
```

Which you can find on the original patch:
https://patch-diff.githubusercontent.com/raw/rails/rails/pull/13972.patch

Thanks!

**Comment by zzak** [comment]
@spencerneste Can you amend the commit message and remove the extra commit data from git? Also for the co-author to work there can't be any whitespace/tabs before the line. LMK if you need help! I think I can commit directly to this branch if necessary :pray:

**Comment by spencerneste** [comment]
@zzak Anything else needed here ? Thanks!

**Comment by spencerneste** [comment]
@rafaelfranca @zzak Mind taking a look ?

File: activerecord/lib/rails/generators/active_record/model/model_generator.rb [hunk 1]
```diff
@@ -58,6 +60,16 @@ def parent_class_name
           end
         end
 
+        def ensure_migration_source_paths
```
Comment (rafaelfranca) [inline activerecord/lib/rails/generators/active_record/model/model_generator.rb:63]: This is pretty much reimplementing [`find_in_source_paths`](https://github.com/rails/thor/blob/376e141adb594f3146c57e98151b97a20c93c484/lib/thor/actions.rb#L133).

Can't we just use it?

Comment (spencerneste) [reply activerecord/lib/rails/generators/active_record/model/model_generator.rb:63]: Thanks for the review @rafaelfranca ! `find_in_source_paths` is being called by [migration_template](https://github.com/rails/rails/blob/main/railties/lib/rails/generators/migration.rb#L57) so we are using it. 

The issue this PR is solving is that `source_paths` do not contain the correct paths for customizable `create_table_migration.rb.tt` templates. So the solution is to include the template_paths from the migration generator, as well a "special" path to preserve existing functionality. 

I'm open to any feedback to get this functionality fixed 😃

Comment (rafaelfranca) [reply activerecord/lib/rails/generators/active_record/model/model_generator.rb:63]: [`Rails::Generators.templates_path` should already be in the source_paths](https://github.com/rails/rails/blob/99096b88ae7a7c9019f2cbcf867e29ba42768e65/railties/lib/rails/generators/base.rb#L242-L259). We don't do any of this work in other generators and they just work.

I think the source of the problem is that we are using relative path to find the template. What we can do is append the migration template to the source path in this generator and change back the relative path.

**Comment by spencerneste** [comment]
@rafaelfranca @zzak Friendly reminder for review. Thanks!

**Comment by spencerneste** [comment]
@zzak Are you able to give this a review? 🙏

**Comment by spencerneste** [comment]
Mind taking a quick look @rafaelfranca ? Would love to get this functionality restored, thanks !

**Pull request merged**